src/Stratis.Bitcoin.Features.SmartContracts/SmartContractMempoolValidator.cs

```diff
@@ -75,7 +81,22 @@ public override void CheckFee(MempoolValidationContext context)
                 rule.CheckTransaction(context);
             }
 
-            // TODO: Check for a minimum gas price on all mining nodes.
+            CheckMinGasLimit(context);
+        }
+
+        private void CheckMinGasLimit(MempoolValidationContext context)
+        {
+            Transaction transaction = context.Transaction;
+
+            if (!transaction.IsSmartContractExecTransaction())
+                return;
+
+            // We know it has passed SmartContractFormatRule so we can deserialize it easily.
+            TxOut scTxOut = transaction.TryGetSmartContractTxOut();
+            Result<ContractTxData> callDataDeserializationResult = this.callDataSerializer.Deserialize(scTxOut.ScriptPubKey.ToBytes());
```

Comment:
Is there a nice way we can avoid performing the deserialization twice?

Reply:
We actually serialize and deserialize way too many times, it would be nice if we could pass a deserialized objects around - currently tracking with #2401